File: tree-number/TreeNumber.scala
package steshaw

object TreeNumber {

  sealed abstract class Tree[+A] {
    def number(seed: Int): (Tree[(A, Int)], Int) = this match {
      case Leaf(n) => (Leaf((n, seed)), seed+1)
      case Branch(l, r) => {
        val (lt, seed1) = l.number(seed)
        val (rt, seed2) = r.number(seed1)
        (Branch(lt, rt), seed2)
      }
    }
    //def number2(seed:Int) = number(seed)
    val number2 = number _
  }
  final case class Leaf[A](a: A) extends Tree[A]
  final case class Branch[A](left: Tree[A], right: Tree[A]) extends Tree[A]

  def main(args: Array[String]) = {
    def go(tree: Tree[Int]) {
      println(tree)
      println(tree.number(10)._1)
    }
    def go2(tree: Tree[Int]) {
      println(tree)
      println(tree.number2(10)._1)
    }

    val t1 = Branch(Branch(Leaf(2), Leaf(6)), Leaf(9))
    val t2 = Branch(Branch(Leaf(2), Branch(Leaf(4), Leaf(5))), Leaf(9))

    go(t1)
    go(t2)
    println

    go2(t1)
    go2(t2)
  }

}

File: tree-number/tree_number.hs
import Control.Monad.State.Lazy
 
data Tree a = Leaf a | Branch (Tree a) (Tree a) deriving Show
 
number :: Int -> Tree a -> (Tree (a, Int),Int)
number seed (Leaf a) = (Leaf (a, seed), seed + 1)
number seed (Branch left right)
 = let (l, ls) = number seed left
       (r, rs) = number ls right
   in
       (Branch l r, rs)
 
numbers :: Tree a -> State Int (Tree (a, Int))
numbers (Leaf a) = do n <- get
                      modify (+1)
                      return (Leaf (a, n))
 
numbers (Branch l r) = do left <- numbers l
                          right <- numbers r
                          return (Branch left right)
 
initState :: State s s
initState = State (\s -> (s, s))

-- test function
main =
  let t = Branch (Branch (Leaf 1) (Leaf 2)) (Leaf 3) in
  putStrLn $ show $ evalState (numbers t) 10
